content/browser/ml/webnn/dml/gpgmm_d3d12.h

```diff
@@ -0,0 +1,356 @@
+// Copyright 2022 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef INCLUDE_MIN_GPGMM_D3D12_H_
+#define INCLUDE_MIN_GPGMM_D3D12_H_
+
+// GPGMM minimum viable implementation (MVI).
+//
+// GPGMM MVI allows users to leverage GPGMM's portable GMM interface without
+// requiring to build the full GPGMM implementation, allowing for incremental
+// enabling during development.
+//
+// GPGMM MVI specifically,
+// * Is not thread-safe.
+// * Is functionally-equivelent to calling
+// ID3D12Device::CreateCommittedResource.
+// * Does not perform residency management.
+//
+// User should decide to define the following macros:
+// - GPGMM_D3D12_HEADERS_ALREADY_INCLUDED: D3D12 platform headers will be
+// already included before this header and does not need to be re-included.
+// - GPGMM_WINDOWS_HEADERS_ALREADY_INCLUDED: Windows.h will be already included
+// before this header and does not need to be re-included.
+//
+#ifndef GPGMM_D3D12_HEADERS_ALREADY_INCLUDED
+#include <d3d12.h>
+#include <dxgi1_4.h>
+#include <wrl.h>
+#endif
+
+#ifndef GPGMM_WINDOWS_HEADERS_ALREADY_INCLUDED
+#include <windows.h>  // for DEFINE_ENUM_FLAG_OPERATORS
+#endif
+
+#include <functional>
+#include <string>
+
+#include "gpgmm.h"
+
+namespace gpgmm::d3d12 {
+
+class IUnknownImpl : public IUnknown {
+ public:
+  IUnknownImpl();
+  virtual ~IUnknownImpl();
+
+  HRESULT STDMETHODCALLTYPE QueryInterface(REFIID riid,
+                                           void** ppvObject) override;
+  ULONG STDMETHODCALLTYPE AddRef() override;
+  ULONG STDMETHODCALLTYPE Release() override;
+
+ protected:
+  virtual void DeleteThis();
+
+ private:
+  uint64_t mRefCount;
+};
+
+enum RESIDENCY_STATUS {
+  RESIDENCY_UNKNOWN = 0,
+  PENDING_RESIDENCY = 1,
+  CURRENT_RESIDENT = 2,
+};
+
+struct HEAP_INFO {
+  bool IsLocked;
+  RESIDENCY_STATUS Status;
+};
+
+enum HEAPS_FLAGS {
+  HEAPS_FLAG_NONE = 0x0,
+  HEAP_FLAG_ALWAYS_IN_BUDGET = 0x1,
+};
+
+DEFINE_ENUM_FLAG_OPERATORS(HEAPS_FLAGS)
+
+struct HEAP_DESC {
+  uint64_t SizeInBytes;
+  uint64_t Alignment;
+  HEAPS_FLAGS Flags;
+  DXGI_MEMORY_SEGMENT_GROUP MemorySegmentGroup;
+  std::string DebugName;
+};
+
+using CreateHeapFn = std::function<HRESULT(ID3D12Pageable** ppPageableOut)>;
```

Comment:
@bbernhar there are below errors when uploading the patch to chromium upstream, please fix it, thanks.

> ** Presubmit ERRORS: 1 **
> Banned functions were used.
>     content\browser\ml\webnn\dml\gpgmm_d3d12.h:86:
>       std::function is banned. Instead use base::OnceCallback or
>       base::RepeatingCallback, which directly support Chromium's weak
>       pointers, ref counting and more.

Reply:
@egalli Please work with Brandon on this. Thanks!

Reply:
I [moved the code of CreateCommittedResource out of callback](https://chromium-review.googlesource.com/c/chromium/src/+/3915513/5..6/content/browser/ml/webnn/dml/gpgmm_d3d12.cpp#290) for the overall reviewer , you can refine it when landing into chromium.

The MVI is integrated in [patchset 6](https://chromium-review.googlesource.com/c/chromium/src/+/3915513/5..6).

Reply:
@fujunwei Thanks!

FYI, I split the headers from the MVI. You'll need to copy over both now.
* https://github.com/intel/GPGMM/tree/main/include
* https://github.com/intel/GPGMM/tree/main/src/mvi

This will allow us to share the same GMM interfaces: MVI or not.